### PR TITLE
Implementing LootTable

### PR DESCRIPTION
**Pull request by AleksKors**
### What has been done?
- Implemented LootTable.
- Created JavaDoc for LootTable.
- Created simple tests for LootTable.

### CODE REVIEW

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     * @param itemChances a map from {@link ItemRarity} to their chances of type double.
+     * @param itemProviders a collection of item providers
+     * @throws IllegalArgumentException if the sum of the chances is greater than 1, or any chance is negative or greater than 1
+	 * @see ItemRarity
+     */
+	LootTable(Map<ItemRarity, Double> itemChances, final Collection<ItemSPI> itemProviders) {
+		this.itemProviders = itemProviders;
+		double sumChance = 0;
+		for (Map.Entry<ItemRarity, Double> entry : itemChances.entrySet()) {
+			Double chance = entry.getValue();
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:37]: Use primitive variant of `double`

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+public class LootTable {
+	private final Map<ItemRarity, Double> itemChances;
+	private final Collection<ItemSPI> itemProviders;
+
+	/**
+     * Constructs a new LootTable with the given item chances.
+     * The item providers are loaded using the ServiceLoader mechanism.
+     *
+     * @param itemChances a map from {@link ItemRarity} to their chances of type double.
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:17]: I think it should read something like?? 
> A map which maps {@link ItemRarity} to a `chance`. 

Please also specify valid values of for `chance`

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	 * @see Optional
+     */
+	Optional<ItemRarity> chooseRarity() {
+		double random = Math.random();
+		double sum = 0;
+		for (Map.Entry<ItemRarity, Double> entry : this.itemChances.entrySet()) {
+			if (random < entry.getValue() + sum) {
+				return Optional.of(entry.getKey());
+			} else {
+				sum += entry.getValue();
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:60]: Can simplify by moving this out of else clause, and then remove empty else clause completely.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:1]: Please fix indentation by reformatting file in Intellij.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ * If {@link #chooseRarity()} returns an Optional.empty(), this method will also return an Optional.empty().
+ *
+ * @return an Optional containing the chosen item, or an empty Optional if no item was chosen.
+ * @throws Exception if no item providers are available for the chosen rarity.
+ * @see #chooseRarity()
+ * @see Optional
+ */
+	public Optional<CommonItem> getItem() throws Exception {
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isPresent()) {
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:78]: Reverse if statement
```java
if (chosenRarity.isEmpty()) return Optional.empty();
/*
 * Rest of code
 */
```
This make the code easier to read, as majority of it is not indented.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			List<ItemSPI> itemSPIList = new ArrayList<>();
+			this.itemProviders.forEach(itemSPI -> {
+						if (itemSPI.getRarity() == chosenRarity.get()) {
+							itemSPIList.add(itemSPI);
+						}
+					}
+			);
+			if (itemSPIList.isEmpty()) throw new Exception("ItemSPIList is empty");
+			Random random = new Random();
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:87]: Store as private final field.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import com.github.sef24sp4.common.item.CommonItem;
+import com.github.sef24sp4.common.item.ItemRarity;
+import com.github.sef24sp4.common.item.ItemSPI;
+
+import java.util.*;
+
+public class LootTable {
+	private final Map<ItemRarity, Double> itemChances;
+	private final Collection<ItemSPI> itemProviders;
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:11]: Might be relevant to replace this with:
```java
private final Map<ItemRarity, Collection<ItemSPI>> itemProviders;

/*
 * then call in constructor
 * this.itemProviders = generateItemProvidersMap(...);
 */

private static Map<ItemRarity, Collection<ItemSPI>> generateItemProvidersMap(final Collection<ItemSPI> providers) {
	final Map<ItemRarity, Collection<ItemSPI>> map = new HashMap<>();
	for (ItemSPI provider : providers) {
		map.computeIfAbsent(provider.getRarity(), k -> new ArrayList<>()).add(provider);
	}
	return Collections.unmodifiableMap(map);
}
```

Can then simplify `getItem` method. Also greatly reduces computations, as provides are only iterated through once.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isPresent()) {
+			List<ItemSPI> itemSPIList = new ArrayList<>();
+			this.itemProviders.forEach(itemSPI -> {
+						if (itemSPI.getRarity() == chosenRarity.get()) {
+							itemSPIList.add(itemSPI);
+						}
+					}
+			);
+			if (itemSPIList.isEmpty()) throw new Exception("ItemSPIList is empty");
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:86]: Simplify according to chances requested to the `itemProviders` field.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/lootTable/LootTableTest.java
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/lootTable/LootTableTest.java:1]: Might be relevant to test more cases. Specifically also test cases which should return empty `Optional`

**Review by MidnightRocket — CHANGES_REQUESTED** [review]

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java (hunk 6 above)
Comment (AleksKors) [reply CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:11]: Not sure if I implemented it correctly, but all other requests should be fixed :)

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/lootTable/LootTable.java:1]: I have noticed that you have accidentally named the package with a capital `T`. Please rename to use all lowercase. Example could be: `loot.table`, `loot_table` or `loottable`. 
Make sure to use Intellij's refactor function. 
[Naming packages](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html)

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+	private ItemSPI mockItemSPI;
+	private CommonItem commonItem;
+
+	@BeforeEach
+	void setUp() {
+		this.rarityChances = new HashMap<>();
+		this.lootTable = new LootTable(this.rarityChances);
+		this.mockItemSPI = mock(ItemSPI.class);
+		this.commonItem = mock(CommonItem.class);
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:29]: Use `@Mock` annotation instead of manual initialisation when possible.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		LootTable lootTableWithSPI = new LootTable(chances, itemProviders);
+
+		// Call getItem() and assert the result
+		Optional<CommonItem> item = lootTableWithSPI.getItem();
+		assertTrue(item.isEmpty());
+	}
+
+	@Test
+	void testSPIListEmpty() throws Exception {
+		Exception emptyItemSPIList = new Exception("ItemSPIList is empty");
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:73]: Remove redundant variable

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class LootTableTest {
+	private LootTable lootTable;
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:18]: Remove instance field if it cannot generically be initialised in `setUp` method.
Currently its initialisation in `setUp` method is overridden by all tests. 
Consider if a generic initialisation method is possible. Otherwise use local variable instead.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+	@Test
+	void testConstructorWithInvalidArguments() {
+		this.rarityChances.put(ItemRarity.COMMON, -0.1);
+		assertThrows(IllegalArgumentException.class, () -> new LootTable(this.rarityChances));
+		this.rarityChances.put(ItemRarity.UNCOMMON, 1.1);
+		assertThrows(IllegalArgumentException.class, () -> new LootTable(this.rarityChances));
+		this.rarityChances.put(ItemRarity.RARE, 0.6);
+		assertThrows(IllegalArgumentException.class, () -> new LootTable(this.rarityChances));
+	}
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:40]: Subsequent tests are currently being "shadowed" by the first test case. 
In other words `IllegalArgumentException` is not necessarily thrown because of the new test case, but because the map is polluted with an illegal state from the previous test. 

This is best illustrated with what is going on line 38-39. 
In an isolated environment this should not throw an exception. However, an exception is still thrown because the map still contains illegal values. 

Consider using [Parameterized tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).

Something like: 
```java
/**
 * Argument source for {@link #testConstructorWithInvalidArguments(Map<ItemRarity, Double>)}.
 *
 * @return The test arguments.
 * @see ParameterizedTest
 * @see MethodSource
 */
public static Stream<Arguments> testConstructorWithInvalidArguments() {
	return Stream.of(
			Arguments.of(Map.ofEntries( // TEST CASE 1
					Map.entry(ItemRarity.COMMON, 1.2)
			)),
			Arguments.of(Map.ofEntries( // TEST CASE 2
					Map.entry(ItemRarity.COMMON, 0.5),
					Map.entry(ItemRarity.RARE, -0.1)
			)),
			Arguments.of(Map.ofEntries( // TEST CASE 4
					Map.entry(ItemRarity.UNCOMMON, 0.2),
					Map.entry(ItemRarity.COMMON, 0.5),
					Map.entry(ItemRarity.RARE, 0.4)
			))
	);
}

@ParameterizedTest
@MethodSource
public void testConstructorWithInvalidArguments(final Map<ItemRarity, Double> map) {
	// ... code
}
```

Consider also making a parameterized test for cases where the constructor should not fail. Can use the `assertDoesNotThrow`. 

> [!TIP]
> Real world **Example**
> https://github.com/SE-F24-SP4/semesterproject-4/blob/2739d551cb1c0fe6bb9c31c71f521408c7ec0436/CollisionSystem/src/test/java/com/github/sef24sp4/collisionsystem/map/BucketTest.java#L222-L227
> Example of the corresponding method source. Where the arguments are provided. 
> https://github.com/SE-F24-SP4/semesterproject-4/blob/2739d551cb1c0fe6bb9c31c71f521408c7ec0436/CollisionSystem/src/test/java/com/github/sef24sp4/collisionsystem/map/BucketTest.java#L81-L97

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	 */
+	public Optional<CommonItem> getItem() throws Exception {
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isEmpty()) return Optional.empty();
+		List<ItemSPI> itemSPIList = new ArrayList<>();
+		Map<ItemRarity, Collection<ItemSPI>> provider = generateItemProvidersMap(this.itemProviders);
+		provider.forEach((itemRarity, itemSPIS) -> {
+			if (itemRarity.equals(chosenRarity.get())) itemSPIList.addAll(itemSPIS);
+		});
+		if (itemSPIList.isEmpty()) throw this.emptyItemSPIList;
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:93]: Instead of throwning an exception consider returning an empty `Optional` instead.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+	@Test
+	void emptyChooseRarity() {
+		Map<ItemRarity, Double> chances = new HashMap<>();
+		chances.put(ItemRarity.COMMON, 0.0);
+		this.lootTable = new LootTable(chances);
+		assertEquals(this.lootTable.chooseRarity(), Optional.empty());
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:47]: Instead of doing a comparison with an empty `Optional` consider asserting the truthfulness of `optional.isEmpty()` method.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+		Map<ItemRarity, Double> chances = new HashMap<>();
+		chances.put(ItemRarity.COMMON, 1.0);
+		this.lootTable = new LootTable(chances);
+		assertEquals(this.lootTable.chooseRarity(), Optional.of(ItemRarity.COMMON));
+	}
+	@Test
+	void testGetEmptyItem() throws Exception {
+		//Create LootTable with ItemSPI
+		List<ItemSPI> itemProviders = new ArrayList<>();
+		itemProviders.add(this.mockItemSPI);
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:61]: Does this test require you to pass in mocked item providers? It does not seem so to me.

**Review by MidnightRocket — CHANGES_REQUESTED** [review]

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java (hunk 12 above)
Comment (AleksKors) [reply CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:93]: It's more that there should be an SPI if chosenRarity isn't empty

Comment (MidnightRocket) [reply CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:93]: The problem is what is the best course of action when such a condition arises? 
Should the game just crash, if there is not a specific item for a given `ItemRarity`?

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+	LootTable(Map<ItemRarity, Double> itemChances, final Collection<ItemSPI> itemProviders) {
+		this.providers = generateItemProvidersMap(itemProviders);
+		double sumChance = 0;
+		for (Map.Entry<ItemRarity, Double> entry : itemChances.entrySet()) {
+			double chance = entry.getValue();
+			if (chance < 0) throw new IllegalArgumentException("The chance of " + entry.getKey() + " is negative");
+			if (chance > 1) throw new IllegalArgumentException("The chance of " + entry.getKey() + " is greater than 1");
+			sumChance += chance;
+		}
+		if (sumChance > 1) throw new IllegalArgumentException("The sum of " + itemChances.keySet() + " is greater than 1");
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:49]: Please extract this to a private static helper method, isolate the validation logic from general constructor logic.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+	 * @return an Optional containing the chosen item, or an empty Optional if no item was chosen.
+	 * @throws Exception if no item providers are available for the chosen rarity.
+	 * @see #chooseRarity()
+	 * @see Optional
+	 */
+	public Optional<CommonItem> getItem() throws Exception {
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isEmpty() || this.providers.get(chosenRarity.get()) == null) return Optional.empty();
+
+		List<ItemSPI> itemSPIList = this.providers.get(chosenRarity.get()).stream().toList();
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:84]: No reason to repack result into list. Just use the raw `Collection` returned.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+	 * Gets an item from the loot table.
+	 * The rarity of the item is chosen using the item chances, and then an item of that rarity is chosen from the item providers.
+	 * If {@link #chooseRarity()} returns an Optional.empty(), this method will also return an Optional.empty().
+	 *
+	 * @return an Optional containing the chosen item, or an empty Optional if no item was chosen.
+	 * @throws Exception if no item providers are available for the chosen rarity.
+	 * @see #chooseRarity()
+	 * @see Optional
+	 */
+	public Optional<CommonItem> getItem() throws Exception {
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:80]: No exception is ever thrown within this method. Remove unnecessary declared checked exception.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+	@ParameterizedTest
+	@MethodSource
+	void testConstructorWithInvalidArguments(final Map<ItemRarity, Double> map) {
+		assertThrows(IllegalArgumentException.class, () -> new LootTable(map));
+	}
+
+	@Test
+	void validConstructor() {
+		assertDoesNotThrow(() -> new LootTable(this.rarityChances));
+	}
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:70]: Could be relevant to make a parameterized test with different test cases.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+		LootTable lootTable = new LootTable(chances);
+		assertTrue(lootTable.chooseRarity().isEmpty());
+	}
+
+	@Test
+	void chooseRarity() {
+		Map<ItemRarity, Double> chances = new HashMap<>();
+		chances.put(ItemRarity.COMMON, 1.0);
+		LootTable lootTable = new LootTable(chances);
+		lootTable = new LootTable(chances);
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:85]: Please remove redundant reinitialisation.

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+	void testGetEmptyItem() throws Exception {
+		//Create LootTable
+		Map<ItemRarity, Double> chances = new HashMap<>();
+		chances.put(ItemRarity.COMMON, 0.0);
+		LootTable lootTable = new LootTable(chances);
+
+		// Call getItem() and assert the result
+		Optional<CommonItem> item = lootTable.getItem();
+		assertTrue(item.isEmpty());
+	}
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:98]: How is this test different from the test below?

File: CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class LootTableTest {
+	@Mock
+	private ItemSPI mockItemSPI;
+	@Mock
+	private CommonItem commonItem;
```
Comment (MidnightRocket) [inline CommonItem/src/test/java/com/github/sef24sp4/common/item/loottable/LootTableTest.java:26]: These fields are only used in one test method. Consider moving to local variables instead.

**Review by MidnightRocket — CHANGES_REQUESTED** [review]

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+		Optional<ItemRarity> chosenRarity = this.chooseRarity();
+		if (chosenRarity.isEmpty() || this.providers.get(chosenRarity.get()) == null) return Optional.empty();
+
+		Collection<ItemSPI> itemSPIs = this.providers.get(chosenRarity.get());
+		int randomInt = new Random().nextInt(itemSPIs.size());
+		Iterator<ItemSPI> iterator = itemSPIs.iterator();
+		for (int i = 0; i < randomInt; i++) {
+			iterator.next();
+		}
+		return Optional.ofNullable(iterator.next().getItem());
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:107]: I apologize, I have made a mistake. I did not realize that `Collection` does not provide a `get(int)` method. 

Please change in relevant places from type `Collection<ItemSPI>` to type `List<ItemSPI>`, so this can be simplified again. 


The ideal is to rely on the simplest possible types which still provides the methods which are required. And I did not realize that `get(int)` was not provided in the `Collection` type.

Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:107]: I am not sure the `ofNullable(T)` is the appropriate method to use. 

If `getItem()` returns `null`, then I would consider that a breach of contract, and should be attributed to programming error, which should not fail silently. 

Please reevaluate and use your own judgment. 
You are very welcomed to comment with your thoughts.

**Review by MidnightRocket — CHANGES_REQUESTED** [review]
It is almost perfect. Unfortunately I have made a mistake in some of my previous feedback, which needs correcting. 
I apologize for that.

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java (hunk 22 above)
Comment (MidnightRocket) [reply CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:107]: Yes but that case is handled here
https://github.com/SE-F24-SP4/semesterproject-4/blob/5055f45e7e276cb26b38510b254c63d6ad463e75/CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java#L99

I am specifically talking about what is going on line [107](https://github.com/SE-F24-SP4/semesterproject-4/blob/5055f45e7e276cb26b38510b254c63d6ad463e75/CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java#L107)

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+	/**
+	 * Gets an item from the loot table.
+	 * The rarity of the item is chosen using the item chances, and then an item of that rarity is chosen from the item providers.
+	 * If {@link #chooseRarity()} returns an Optional.empty() or there isn't an ItemSPI with the rarity of the {@link #chooseRarity()}, this method will also return null.
+	 *
+	 * @return Returns an item or null if no item was chosen.
+	 * @see #chooseRarity()
+	 */
+	public CommonItem getItem() {
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:95]: This method should definitely still return an `Optional`

**Review by MidnightRocket — CHANGES_REQUESTED** [review]

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+import com.github.sef24sp4.common.item.CommonItem;
+import com.github.sef24sp4.common.item.ItemRarity;
+import com.github.sef24sp4.common.item.ItemSPI;
+
+import java.util.*;
+
+public class LootTable {
+	private final Map<ItemRarity, Double> itemChances;
+	private final Map<ItemRarity, Collection<ItemSPI>> providers;
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:11]: Change this to type `Map<ItemRarity, Liste<ItemSPI>>`

File: CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+
+	/**
+	 * Generates a map from {@link ItemRarity} to a collection of {@link ItemSPI} instances.
+	 *
+	 * @param providers A collection of item providers.
+	 * @return A map from {@link ItemRarity} to a collection of {@link ItemSPI} instances.
+	 */
+	private static Map<ItemRarity, Collection<ItemSPI>> generateItemProvidersMap(final Collection<ItemSPI> providers) {
+		final Map<ItemRarity, Collection<ItemSPI>> map = new HashMap<>();
```
Comment (MidnightRocket) [inline CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java:20]: Change from `Map<ItemRarity, Collection<ItemSPI>>` to type `Map<ItemRarity, Liste<ItemSPI>>`

**Review by MidnightRocket — APPROVED** [review]
